test/rules/use_enums_test.dart

```diff
@@ -0,0 +1,387 @@
+    await assertDiagnostics(r'''
+class _E {
+  static const _E a = _E();
+  static const _E b = _E();
+
+  const _E({_E e = const _E()});
+}
+''', [
+      // No lint.
+      error(HintCode.UNUSED_ELEMENT, 6, 2),
+      error(HintCode.UNUSED_FIELD, 29, 1),
+      error(HintCode.UNUSED_FIELD, 57, 1),
+      error(HintCode.UNUSED_ELEMENT_PARAMETER, 83, 1),
+    ]);
+  }
+
+  test_no_lint_constructorUsedOutsideClass() async {
+    await assertDiagnostics(r'''
+class _E {
+  static const _E a = _E();
+  static const _E b = _E();
+
+  const _E();
+}
+
+_E get e => _E();
+''', [
+      // No lint.
+      error(HintCode.UNUSED_FIELD, 29, 1),
+      error(HintCode.UNUSED_FIELD, 57, 1),
+    ]);
+  }
+
+  test_no_lint_extended() async {
+    await assertNoDiagnostics(r'''
+class A {
+  static const A a = A._();
+  static const A b = A._();
+  const A._();
+}
+class B extends A {
+  const B() : super._();
+}
+''');
+  }
+
+  test_no_lint_extends_notObject() async {
+    await assertNoDiagnostics(r'''
+class O { 
+  const O();
+}    
+class A extends O {
+  static const A a = A._();
+  static const A b = A._();
+  const A._();
+}
+''');
+  }
+
+  test_no_lint_factoryAll() async {
+    await assertDiagnostics(r'''
+class _E {
+  static _E c = _E();
+  static _E d = _E();
+
+  factory _E() => c;
+}
+''', [
+      // No lint.
+      error(HintCode.UNUSED_ELEMENT, 6, 2),
+      error(HintCode.UNUSED_FIELD, 45, 1),
+    ]);
+  }
+
+  test_no_lint_factorySome() async {
+    await assertDiagnostics(r'''
+class _E {
+  static _E c0 = _E._();
+  static _E c1 = _E();
+
+  factory _E() => c0;
+  const _E._();
+}
+''', [
+      // No lint.
+      error(HintCode.UNUSED_ELEMENT, 6, 2),
+      error(HintCode.UNUSED_FIELD, 48, 2),
+    ]);
+  }
+
+  test_no_lint_implemented() async {
+    await assertDiagnostics('''
+class _E {
+  static const _E c = _E();
+  static const _E d = _E();
+
+  const _E();
+}
+class F implements _E  {}
+''', [
+      // No lint.
+      error(HintCode.UNUSED_FIELD, 29, 1),
+      error(HintCode.UNUSED_FIELD, 57, 1),
+    ]);
+  }
+
+  test_no_lint_implements_index_field() async {
+    await assertNoDiagnostics(r'''
+class A {
+  static const A a = A._();
+  static const A b = A._();
+  const A._();
+  final int index = 0;
+}
+''');
+  }
+
+  test_no_lint_implements_index_getter() async {
+    await assertNoDiagnostics(r'''
+class A {
+  static const A a = A._();
+  static const A b = A._();
+  const A._();
+  int get index => 0;
+}
+''');
+  }
+
+  test_no_lint_implements_values_field() async {
+    await assertNoDiagnostics(r'''
+class A {
+  static const A a = A._();
+  static const A b = A._();
+  const A._();
+  static final List<A> values = [a, b];
+}
+''');
+  }
+
+  test_no_lint_implements_values_method() async {
+    await assertNoDiagnostics(r'''
+class A {
+  static const A a = A._();
+  static const A b = A._();
+  const A._();
+  static List<A> values() => [];
+}
+''');
+  }
+
+  test_no_lint_nonConstConstructor() async {
```

Comment:
This is another case that we might want to relax in the lint. If there's a non-const constructor but all of the fields are final, then it might still be reasonable to flag the class.